src/Core/CustomObjects/CustomProperty.cs

```diff
@@ -10,54 +10,67 @@ namespace Axe.Windows.Core.CustomObjects
+
+        private string _ProgrammaticName;
         /// <summary>A textual description of this property.</summary>
         [JsonProperty("programmaticName")]
-        public string ProgrammaticName { get; set; }
-
-        /// <summary>The data type of this property's value as specified by the user, one of string, int, bool, double, point, or element.</summary>
-        // TODO Bill: add enum (with values member)
-        [JsonProperty("uiaType")]
-        public string DataType { get; set; }
-
-        /// <summary>A type converter for this property, providing string rendering.</summary>
-        [JsonIgnore]
-        public ITypeConverter TypeConverter { get; private set; }
-
-        /// <summary>The dynamic ID assigned to this property by the system.</summary>
-        [JsonIgnore]
-        public int DynamicId { get; set; }
-
-        internal void Validate()
-        {
-            if (Guid == Guid.Empty) throw new InvalidDataException("Missing GUID in custom property definition.");
-            if (ProgrammaticName == null) throw new InvalidDataException("Missing programmatic name in custom property definition.");
-            if (DataType == null) throw new InvalidDataException("Missing type in custom property definition.");
-            TypeConverter = CreateTypeConverter();
+        public string ProgrammaticName {
+            get { return _ProgrammaticName; }
+            set
+            {
+                if (value == null) throw new InvalidDataException("Missing programmatic name in custom property definition.");
+                _ProgrammaticName = value;
+            }
         }
 
-        internal ITypeConverter CreateTypeConverter()
+        ///  <summary>An internal representation of this property's type. For performance reasons, calling code should always use this property in place of the config representation.s</summary>
+        public CustomUIAPropertyType Type { get; private set; }
+
+        private string _ConfigType;
+        /// <summary>The data type of this property's value as specified in user configuration, one of string, int, bool, double, point, or element.</summary>
+        [JsonProperty("uiaType")]
+        public string ConfigType
         {
-            switch (DataType)
+            get { return _ConfigType; }
+            set
             {
-                case "string":
-                    return new StringTypeConverter();
-                case "int":
-                    return new IntTypeConverter();
-                case "bool":
-                    return new BoolTypeConverter();
-                case "double":
-                    return new DoubleTypeConverter();
-                case "point":
-                    return new PointTypeConverter();
-                case "element":
-                    return new ElementTypeConverter();
-                default:
-                    throw new InvalidDataException($"Unknown type {DataType}.");
+                switch (value)
```

Comment:
We should include unit tests to pin the behavior of each of these possible branches